### PR TITLE
Explain how StrictMode affects function components and hooks

### PR DESCRIPTION
**Pull request by eps1lon**
The StrictMode docs were not updated with the release of hooks and it isn't obvious which hooks are invoked twice.

Codesandbox for the new cases: https://codesandbox.io/s/strictmode-double-invokation-xdpxj

Encountered this question in
- https://github.com/mui-org/material-ui/issues/18083#issue-513863598
- https://github.com/mui-org/material-ui/issues/20313
- somewhere else but I can't remember where

Preview: https://deploy-preview-2515--reactjs.netlify.com/docs/strict-mode.html#detecting-unexpected-side-effects

### CODE REVIEW

File: content/docs/strict-mode.md [hunk 1]
```diff
@@ -100,9 +100,12 @@ Because the above methods might be called more than once, it's important that th
 Strict mode can't automatically detect side effects for you, but it can help you spot them by making them a little more deterministic. This is done by intentionally double-invoking the following methods:
 
 * Class component `constructor` method
-* The `render` method
-* `setState` updater functions (the first argument)
+* The `render` method of class components or the function component itself 
+* state updater functions (the first argument for the `setState` class component method or the argument for the `setState` return value of [`useState`](hooks-reference.html#usestate))
 * The static `getDerivedStateFromProps` lifecycle
+* the function passed as the initial state to [`useState`](hooks-reference.html#lazy-initial-state)
+* the `init` function passed to [`useReducer`](/hooks-reference.html#usereducer)
```
Comment (eps1lon) [inline content/docs/strict-mode.md]:
```suggestion
* the `init` function passed to [`useReducer`](hooks-reference.html#usereducer)
```

File: content/docs/strict-mode.md [hunk 2]
```diff
@@ -100,9 +100,12 @@ Because the above methods might be called more than once, it's important that th
...
+* "create" functions passed to [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo)
```
Comment (eps1lon) [inline content/docs/strict-mode.md]:
```suggestion
* "create" functions passed to [`useMemo`](hooks-reference.html#usememo)
```

**Comment by gaearon** [comment]
Let's remove "only if it uses Hooks". This really isn't an important detail, and as you sent a PR, we'll change that.

File: content/docs/strict-mode.md [hunk 3]
```diff
@@ -100,10 +100,13 @@ Because the above methods might be called more than once, it's important that th
 Strict mode can't automatically detect side effects for you, but it can help you spot them by making them a little more deterministic. This is done by intentionally double-invoking the following methods:
 
 * Class component `constructor` method
-* The `render` method
-* `setState` updater functions (the first argument)
+* The `render` method of class components or the function component itself **only if** it uses [Hooks](hooks-intro.html)
```
Comment (eps1lon) [inline content/docs/strict-mode.md]:
```suggestion
* The `render` method of class components or the function component itself
```

**Comment by gaearon** [comment]
I tweaked the explanation a bit to remove distracting details and focus on the essence.

**Comment by eps1lon** [comment]
> I tweaked the explanation a bit to remove distracting details and focus on the essence.

That is much better, thanks!

**Pull request merged**